### PR TITLE
Implement State Store for OCI ObjectStorage service

### PR DESCRIPTION
**Pull request by lucasjellema**
# Description

The OCI Object Storage Service is a cloud service on Oracle Cloud Infrastructure, very similar in nature to AWS S3 and Azure BlobStorage. This services support creating, retrieving, deleting objects that represent state. It has concurrency control through ETags. I believe that especially for Daprized applications running on OCI – either on OKE, the OCI container engine for Kubernetes or on a compute instance – it can be of value to leverage ObjectStorage as the state store. It is cheap, easily accessible and can easily be integrated into many technologies from within the OCI cloud as well as from elsewhere.

The contribution allows Dapr application operators to configure state store components backed by OCI Object Storage.

The code I submit in this PR has been described in some detail in [an article on Medium](https://lucasjellema.medium.com/composing-dapr-custom-state-store-component-for-oracle-cloud-object-storage-a4c2feb15b3). 

I am not sure how to deal with the conformance tests. If these are required as this point, please provide some guidance as to how to create these.

I have resolved most Linting alerts. There are some left of type exhaustivestruct. I fear that resolving these makes the code less readable and certainly less elegant. If you insist, I can of course fix these too. 

Note: the Dapr API for Set does not include a SetResponse. That means in the case of OCI Object Storage (that assigns generated ETag values) that the ETag assigned after a Set has been performed cannot be returned. Only by doing a Get immediately after a Set will the application get the ETag value - and not even guaranteed the correct one as potentially another process can already have changed the state. It would be nice to have a way to return this ETag from the Set operation.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[1400]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[2070]_  [link](https://github.com/dapr/docs/issues/2070)

### CODE REVIEW

**Comment by jjcollinge** [comment]
Just a quick question on your NOTE about etag in the set response. I agree returning the etag in the set response could be an optimization (1 call vs 2 calls) but is there a behavioural reason you need the exact etag from your write? AFAIK, you cannot guarantee you have the latest etag until you try to write again so adding the etag to th set response wouldn't give you any additional guarantees other than this is the etag from your write (which may already be stale).

**Comment by lucasjellema** [comment]
> 



> Just a quick question on your NOTE about etag in the set response. I agree returning the etag in the set response could be an optimization (1 call vs 2 calls) but is there a behavioural reason you need the exact etag from your write? AFAIK, you cannot guarantee you have the latest ("correct") etag until you try to write again so adding the etag to the set response wouldn't give you any additional guarantees other than this is the etag from your write (which may already be stale).

I think you are right. The only use case where having that freshly assigned etag at your disposal is when you intend to do multiple updates of the same state element - which is very unlikely. Perhaps in case an in memory state store (cache) is used with a second persisted state store as backup where the first level cache contains both etag and key & value in order to process state changes... It sounds quite far fetched right now. So perhaps there is not much case to be made after all for returning the etag in a Set response.

**Comment by lucasjellema** [comment]
I am new at contributing to the Dapr project. I am not sure what to do next. Should I ask for the workflows to be kicked off again? Should I request another round of review? How can I do that?

**Comment by yaron2** [comment]
> I am new at contributing to the Dapr project. I believe the code is now at a level where it is ready for review. I am not sure what to do next. Should I ask for the workflows to be kicked off again? Should I request another round of review? How can I do that?

An approver or maintainer will review the PR. We have a bit of a queue of PRs on contrib at the moment.

File: state/oci/objectstorage/objectstorage.go
Comment (daixiang0) [inline state/oci/objectstorage/objectstorage.go:1]: Need to add apache license header, same as below, refer to other files please.

Comment (daixiang0) [reply state/oci/objectstorage/objectstorage.go:1]: Those helpful info should comment in the doc, not here I think.

File: state/oci/objectstorage/objectstorage_test.go
Comment (daixiang0) [inline state/oci/objectstorage/objectstorage_test.go:3]: Remove those 3 lines.

File: state/oci/objectstorage/objectstorage_test.go
Comment (daixiang0) [inline state/oci/objectstorage/objectstorage_test.go:19]: Please add more cases for config to ensure any wrong config would return error.

**Review by daixiang0 — COMMENTED** [review]
Thanks for your contribution! Please fix CI issues and see comments above.

**Comment by daixiang0** [comment]
Will fix #1400

File: state/oci/objectstorage/objectstorage_test.go
Comment (lucasjellema) [reply state/oci/objectstorage/objectstorage_test.go]: I have included test cases for each of the configuration properties - all are mandatory at this point.

**Comment by lucasjellema** [comment]
I believe I have satisfied the linting requirements as well have implemented the per review requested changes.

**Comment by lucasjellema** [comment]
The PR is now failing on two checks. One concerns file tests/certification/pubsub/kafka/go.sum and the other tests/certification/pubsub/rabbitmq/go.sum. I am not sure how these issues popped up in my PR - considering that I did not touch these files. 

Can you provide guidance as to how I can proceed? Note: I noticed that https://github.com/dapr/components-contrib/pull/1362 and https://github.com/dapr/components-contrib/pull/1402 are failing for the same reason. Could it perhaps be an issue with the check rather than the PR?

**Review by daixiang0 — APPROVED** [review]
Codes look good to me, wait for CI.

**Comment by yaron2** [comment]
@lucasjellema please run `go mod tidy` and push the changes.

**Comment by lucasjellema** [comment]
> @lucasjellema please run `go mod tidy` and push the changes.

I have pulled all changes, executed `go mod tidy` several times but I do not get any changes to go.mod.  My version of go.mod has all changes from the master - with two entries added because of the OCI Go SDK. I am not sure what more I can do. Do you have other suggestions?  What is lacking from/incorrect in the go.mod in my PR?

**Comment by yaron2** [comment]
> > @lucasjellema please run `go mod tidy` and push the changes.
> 
> I have pulled all changes, executed `go mod tidy` several times but I do not get any changes to go.mod. My version of go.mod has all changes from the master - with two entries added because of the OCI Go SDK. I am not sure what more I can do. Do you have other suggestions? What is lacking from/incorrect in the go.mod in my PR?

Our certification tests are failing due to mod changes. Are you using Go 1.17?

**Comment by lucasjellema** [comment]
> > > @lucasjellema please run `go mod tidy` and push the changes.
> > 
> > 
> > I have pulled all changes, executed `go mod tidy` several times but I do not get any changes to go.mod. My version of go.mod has all changes from the master - with two entries added because of the OCI Go SDK. I am not sure what more I can do. Do you have other suggestions? What is lacking from/incorrect in the go.mod in my PR?
> 
> Our certification tests are failing due to mod changes. Are you using Go 1.17?

I am using Go 1.17.5 on Ubuntu (Linux/amd64)

I noticed that the failing tests are also failing in two other recently updated PRs. I was wondering if perhaps something has changed recently in the tests themselves - that might be causing the issue.

**Comment by yaron2** [comment]
> > > @lucasjellema please run `go mod tidy` and push the changes.
> > 
> > 
> > I have pulled all changes, executed `go mod tidy` several times but I do not get any changes to go.mod. My version of go.mod has all changes from the master - with two entries added because of the OCI Go SDK. I am not sure what more I can do. Do you have other suggestions? What is lacking from/incorrect in the go.mod in my PR?
> 
> Our certification tests are failing due to mod changes. Are you using Go 1.17?

Actually, it looks like these tests are broken for some time. No action needed.

**Comment by yaron2** [comment]
Thanks for your contribution @lucasjellema and welcome to the Dapr contributor community. Please proceed with am additional docs PR, and another PR in the dapr/Dapr repository to register the component.

**Comment by lucasjellema** [comment]
Cool! Thanks for your help. Very nice to reach this stage. 
I have created the PR for Docs yesterday - it is currently pending.
I will now work on the PR for main repository.

It is a great experience to become part of this community. I enjoy your careful and considerate communication style. And I would also like to say that by and large the online instructions for creating and submitting contributions are very clear. The process that takes place once the PR is submitted - automated workflow, role of reviewers - was a little less clear. I have been following the progress of the PR in anxious anticipation. Submitting a PR feels a little bit like doing an exam. And now it feels like I have passed it.

**Comment by yaron2** [comment]
> Cool! Thanks for your help. Very nice to reach this stage. I have created the PR for Docs yesterday - it is currently pending. I will now work on the PR for main repository.
> 
> It is a great experience to become part of this community. I enjoy your careful and considerate communication style. And I would also like to say that by and large the online instructions for creating and submitting contributions are very clear. The process that takes place once the PR is submitted - automated workflow, role of reviewers - was a little less clear. I have been following the progress of the PR in anxious anticipation. Submitting a PR feels a little bit like doing an exam. And now it feels like I have passed it.

Haha, yes there are certain check boxes that we need to tick. Your PR was reviewed very quickly compared to others, hopefully we can onboard new approvers and maintainers to streamline this process further.

**Comment by lucasjellema** [comment]
I am working on adding the component in daprd in the main Dapr repo. I am not sure I understand the steps lighlighted. 
image
When I run `go get -u github.com/dapr/components-contrib@master` I get the code tag v1.5.1 which does not include my PR and has a go.mod without the OCI SDK import that my component requires. Compilation of the Dapr application fails because of the missing import. (the instruction is also unclear as to where to run go modtidy-all and why that should be done). The statement _Ensure that component-contrib is updated to the latest version_ is also somewhat unclear to me: where should that be done (locally I guess) and what does the _latest version_ mean (I would expect that to be some version that includes my approved and merged PR on components-contrib repo?

These instructions do not mandate to locally compile and lint main.go for Daprd . Is that on purpose? If so, it would be useful to explicitly state that fact. Can I just make the code change and submit the PR and have the automated Git actions take over linting/compilation etcetera?

**Comment by yaron2** [comment]
> I am working on adding the component in daprd in the main Dapr repo. I am not sure I understand the steps lighlighted. image When I run `go get -u github.com/dapr/components-contrib@master` I get the code tag v1.5.1 which does not include my PR and has a go.mod without the OCI SDK import that my component requires. Compilation of the Dapr application fails because of the missing import. (the instruction is also unclear as to where to run go modtidy-all and why that should be done). The statement _Ensure that component-contrib is updated to the latest version_ is also somewhat unclear to me: where should that be done (locally I guess) and what does the _latest version_ mean (I would expect that to be some version that includes my approved and merged PR on components-contrib repo?
> 
> These instructions do not mandate to locally compile and lint main.go for Daprd . Is that on purpose? If so, it would be useful to explicitly state that fact. Can I just make the code change and submit the PR and have the automated Git actions take over linting/compilation etcetera?

Run `go get -u github.com/dapr/components-contrib@master` and then `go mod tidy`.

Make sure that in `daprd/main.go` that package you're importing is `github.com/dapr/components-contrib/state/oci/objectstorage`.

FYI, there's a different for Go between latest and master. latest corresponds to the latest tag, while master in this case points to the last commit of master branch, which is what we need :)

GitHub actions will just validate linting and compilation.

**Pull request merged**
